hermes/src/proof_store.rs

```diff
@@ -0,0 +1,207 @@
+    },
+    serde::{
+        Deserialize,
+        Serialize,
+    },
+    std::{
+        collections::{
+            HashMap,
+            HashSet,
+        },
+        ops::{
+            Deref,
+            DerefMut,
+        },
+    },
+    wormhole::VAA,
+};
+
+
+pub enum ProofUpdate {
+    Vaa(Vec<u8>),
+}
+
+pub enum ProofType {
+    BatchVaa,
+}
+
+// TODO: Is this the right type for the proof?
+#[derive(
+    Clone,
+    Default,
+    PartialEq,
+    PartialOrd,
+    Debug,
+    Serialize,
+    Deserialize,
+    BorshSerialize,
+    BorshDeserialize,
+)]
+pub struct Proof(pub Vec<Vec<u8>>);
+
+impl Deref for Proof {
+    type Target = Vec<Vec<u8>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Proof {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+
+// TODO: We need to add more metadata to this struct.
+#[derive(Clone, Default, PartialEq, Debug, BorshSerialize, BorshDeserialize)]
+pub struct PriceInfo {
+    pub price_feed:   PriceFeed,
+    pub vaa_bytes:    Vec<u8>,
+    pub publish_time: UnixTimestamp,
+}
+
+#[derive(Clone, Default, Serialize, Deserialize)]
+pub struct PriceFeedsWithProof {
+    pub price_feeds: HashMap<PriceIdentifier, PriceFeed>,
+    pub proof:       Proof,
+}
+
+
+#[derive(Clone, Default)]
+pub struct ProofStore<D: Db> {
+    pub db: D,
+}
+
+impl<D: Db> ProofStore<D> {
+    pub fn new(db: D) -> Self {
+        Self { db }
+    }
+
+    // TODO: This should return the updated feeds so the subscribers can be notified.
+    pub fn process_update(&mut self, update: ProofUpdate) -> Result<()> {
+        match update {
+            ProofUpdate::Vaa(vaa_bytes) => {
+                // FIXME: Vaa bytes might not be a valid Pyth BatchUpdate message nor originate from Our emitter. We should check that.
+                let vaa = VAA::from_bytes(&vaa_bytes)?;
+                // Ideally this part would be in a separate function/module when we add more proof types.
+                let batch_price_attestation =
+                    BatchPriceAttestation::deserialize(vaa.payload.as_slice())
+                        .map_err(|_| anyhow!("Failed to deserialize VAA"))?;
+
+                for price_attestation in batch_price_attestation.price_attestations {
+                    let price_feed = price_attestation_to_price_feed(price_attestation);
+
+                    let publish_time = price_feed.get_price_unchecked().publish_time.try_into()?;
+
+                    let price_info = PriceInfo {
+                        price_feed,
+                        vaa_bytes: vaa_bytes.clone(),
+                        publish_time,
+                    };
+
+                    let key = price_feed.id.to_bytes().to_vec();
+                    let record = DbRecord {
+                        time:  publish_time,
+                        value: price_info.try_to_vec()?,
+                    };
+                    self.db.insert(&key, record)?;
```

Comment:
duplicating the proofs for each key (e.g., cloning the vaa) is going to be really memory inefficient once we get the merkle trees. It's going to clone each merkle tree O(price feeds) times. Is there a way you can make all the db entries point to the same piece of memory?